### PR TITLE
Opds categories feed

### LINKED ISSUE #518: OPDS endpoint for categories
In order to improve ZIM files filtering on readers, we need to maintain a centralized list of *Categories* to which ZIM files are attached to. This has been mentionned in [kiwix-lib#130](https://github.com/kiwix/kiwix-lib/issues/130) and [kiwix-desktop#193](https://github.com/kiwix/kiwix-desktop/issues/193).

The Categories list will start with the followings and should support internationalizations (`en` translations bellow).

* `gutenberg`: `Gutenberg`
* `phet`: `PhET`
* `psiram`: `Psiram`
* `stack_exchange`: `Stack Exchange`
* `ted`: `TED`
* `vikidia`: `Vikidia`
* `wikibooks`: `Wikibooks`
* `wikinews`: `Wikinews`
* `wikipedia`: `Wikipedia`
* `wikiquote`: `Wikiquote`
* `wikisource`: `Wikisource`
* `wikispecies`: `Wikispecies`
* `wikiversity`: `Wikiversity`
* `wikivoyage`: `Wikivoyage`
* `wiktionary`: `Wiktionary`

### PR DESCRIPTION
**Pull request by veloman-yunkan**
Will fix #518

This PR introduces a new OPDS API at `/catalog/v2` (as well as fixes a few noticed issues unrelated to that enhancement). 
Currently available endpoints are:

- `/catalog/v2/root.xml`: top level navigation feed, pointing to the full acquisition feed and to the list of categories
- `/catalog/v2/entries`: an equivalent of both `/catalog/root.xml` and `/catalog/search` of the old OPDS API.
- `/vatalog/v2/categories`: a navigation feed containing an entry for each category with a link to the corresponding acquisition feed filtered by that category.

Notes:

- facets are not used (contrary to the suggestion in https://github.com/kiwix/libkiwix/issues/518#issuecomment-792627977)
- `OPDSDumper` is not utilized. `mustache` was used instead.

### CODE REVIEW

**Comment by veloman-yunkan** [comment]
We will not have have clean CI until #522 is merged and #523 is fixed. If possible, please review without waiting for the CI to pass.

File: src/server/internalServer.cpp
Comment (mgautierfr) [inline src/server/internalServer.cpp:793]: This seems pretty close to `search_catalog` method. We should reuse it.

File: src/server/internalServer.cpp
Comment (mgautierfr) [inline src/server/internalServer.cpp:834]: Even if this is not mandatory by opds spec, it would be better to also
add openSearch element (https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md#elements)
(As OPDSDumper does)

---

If you need to search for opensearch spec, it seems that now opensearch.org is use for the opensearch software (:facepalm: :angry: :cry:). So all links you can find on the web are wrong, especially the ones in opds specs. The opensearch spec is now at https://github.com/dewitt/opensearch

**Review by mgautierfr — CHANGES_REQUESTED** [review]
> OPDSDumper is not utilized. mustache was used instead.

I dislike the idea. Now we have two ways to generate opds feed.
I'm not against using mustache, but it would be better to port OPDSDumper to mustache and use OPDSDumper everywhere.
(And having a specific OPDSDumper class would allow us to test it without creating a instance of the server)

> /catalog/v2/entries: an equivalent of both /catalog/root.xml and /catalog/search of the old OPDS API.

With issue https://github.com/kiwix/libkiwix/issues/209 the plan is to move to a partial feed for the entries.
In this PR, the entries feed is a complete acquisition feed.
It is not a blocker for this PR, it is probably better to do it in two PR.
But we must care to not do a release with a complete acquisition feed.

Also, now we have only one url for listing all entries and search for them. I don't know if it is good thing or not, but I prefer to mention it explicitly.

**Comment by kelson42** [comment]
If we change the URL paths/schemas we need to be careful. We have Kiwix Desktop and Kiwix iOS relying on them!

**Comment by mgautierfr** [comment]
> If we change the URL paths/schemas we need to be careful. We have Kiwix Desktop and Kiwix iOS relying on them!

We don't change the paths/schemas, we add new ones (`catalog/v2/*`)

File: src/server/internalServer.cpp
Comment (veloman-yunkan) [reply src/server/internalServer.cpp:92]: Fully agree. I simply wanted to avoid any impact on the old code, but it seems that the chance of breaking something with this is negligible.

File: src/server/internalServer.cpp
Comment (veloman-yunkan) [reply src/server/internalServer.cpp:834]: Done

File: src/server/internalServer.h
Comment (veloman-yunkan) [reply src/server/internalServer.h:79]: Done

**Comment by veloman-yunkan** [comment]
> > OPDSDumper is not utilized. mustache was used instead.
> 
> I dislike the idea. Now we have two ways to generate opds feed.
> I'm not against using mustache, but it would be better to port OPDSDumper to mustache and use OPDSDumper everywhere.
> (And having a specific OPDSDumper class would allow us to test it without creating a instance of the server)

I moved most of the new code into `OPDSDumper` but didn't port pre-existing code to mustache. That can be done in another PR.

**Comment by veloman-yunkan** [comment]
New commits start from [/catalog/v2/entries going through OPDSDumper](60b7e667e571ee87d245590b83e8e773ec505d28)

**Comment by mgautierfr** [comment]
The ideas is not to "have a OPDSDumper" but to avoid having two different implementations to do the same thing (transform a list of books, with some metadata, to a opds stream).
Moving mustache implementation inside OPDSDumper but in another method and keep the previous implementation is useless.
If mustache is a better implementation, then we should replace the previous implementation.

**Comment by veloman-yunkan** [comment]
> The ideas is not to "have a OPDSDumper" but to avoid having two different implementations to do the same thing (transform a list of books, with some metadata, to a opds stream).
> Moving mustache implementation inside OPDSDumper but in another method and keep the previous implementation is useless.
> If mustache is a better implementation, then we should replace the previous implementation.

Done

File: src/opds_dumper.cpp
Comment (mgautierfr) [inline src/opds_dumper.cpp:99]: It seems that this data is pretty closed from the ones used in `dumpOPDSFeedV2`.
Could be have "only one method" using two different templates ?
(Or maybe even better, one template with just the endpoint changing ? If possible)

**Comment by kelson42** [comment]
@veloman-yunkan @mgautierfr This PR is key and need review and rebase.

**Review by mgautierfr — CHANGES_REQUESTED** [review]
Two small commenst on the code to fix and we are good. This is far better with all the code in the rendered and using mustache everywhere.

---

On a more general comment, #209 implies that the root/search feed is partial and then we get the full information doing another request (and so another endpoint).
We will definitively not do it in this PR. But I would like to mention that the v2 API may change in close future and client should not start to use it right now.
(And to have a more stable API, we may move the endpoint `/catalog/v2/entries` to `/catalog/v2/search` (and so keep `entry`/`entries` for the endpoint to get all information about entries)

File: src/opds_dumper.cpp
Comment (veloman-yunkan) [reply src/opds_dumper.cpp:99]: Let's consider that after the new API is finalized. I don't want the old API to be a drag on the new API, and a premature attempt to share code between them will do just that.

**Comment by veloman-yunkan** [comment]
@mgautierfr  @kelson42 Only the last two commits are new since the last review (ignoring a a trivial conflict in one of the earlier commits that was resolved while rebasing).

**Comment by mgautierfr** [comment]
We are good. Please rebase-fixup and we can merge.

**Review by mgautierfr — APPROVED** [review]

**Comment by veloman-yunkan** [comment]
> We are good. Please rebase-fixup and we can merge.

Done

**Pull request merged**
